tools/fp20compiler/rc1.0_final.cpp

```diff
@@ -31,29 +31,30 @@ void FinalProductStruct::ZeroOut()
     e.Init(zero, MAP_UNSIGNED_IDENTITY);
     f.Init(zero, MAP_UNSIGNED_IDENTITY);
 }
-void FinalCombinerStruct::Validate()
-{
-    if (hasProduct &&
-        (REG_E_TIMES_F == product.e.reg.bits.name ||
-         REG_SPARE0_PLUS_SECONDARY_COLOR == product.e.reg.bits.name ||
-         REG_DISCARD == product.e.reg.bits.name ||
-         REG_E_TIMES_F == product.f.reg.bits.name ||
-         REG_SPARE0_PLUS_SECONDARY_COLOR == product.f.reg.bits.name ||
-         REG_DISCARD == product.f.reg.bits.name))
-        errors.set("invalid input register for final_product");
-
-    if (hasProduct &&
-        (RCP_BLUE == product.e.reg.bits.channel ||
-         RCP_BLUE == product.f.reg.bits.channel))
-        errors.set("blue register used in final_product");
 
-    if (REG_E_TIMES_F == alpha.g.reg.bits.name ||
-        REG_SPARE0_PLUS_SECONDARY_COLOR == alpha.g.reg.bits.name ||
-        REG_DISCARD == alpha.g.reg.bits.name)
-        errors.set("invalid input register for final alpha");
+static void validateRgbRegister(MappedRegisterStruct& reg, bool isEorF, bool hasProduct) {
+    line_number = reg.line_number;
+    if (REG_DISCARD == reg.reg.bits.name)
+        errors.set("reading from discard", line_number); //FIXME: Be more explicit
+    if (RCP_BLUE == reg.reg.bits.channel)
+        errors.set("blue register used in final rgb", line_number);
+    if (!isEorF) {
+        if (!hasProduct && (REG_E_TIMES_F == reg.reg.bits.name))
+            errors.set("final_product used but not set", line_number);
+    } else {
+        if (REG_E_TIMES_F == reg.reg.bits.name)
+            errors.set("final_product can not reference itself", line_number);
+        if (REG_SPARE0_PLUS_SECONDARY_COLOR == reg.reg.bits.name)
+            errors.set("invalid input register for final_product", line_number);
+    }
 
-    if (RCP_RGB == alpha.g.reg.bits.channel)
-        errors.set("rgb register used in final alpha");
+    if (RCP_NONE == reg.reg.bits.channel)
+        reg.reg.bits.channel = RCP_RGB;
+}
+
+void FinalCombinerStruct::Validate()
+{
+    int line_number;

```

Comment:
This variable can be removed now. It's far less used than before.

Reply:
It's still being used..
I guess I could remove it if I also move `alpha.g.reg` checks into a validation function (only 1 use would remain, which can be inlined).